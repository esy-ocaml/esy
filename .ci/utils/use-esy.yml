# steps to install esy globally

steps:
- script: "yarn global add @esy-nightly/esy@0.6.8-bbdfb8"
  displayName: "Install esy (via yarn)"
  condition: ne(variables['AGENT.OS'], 'Windows_NT')
  
- script: |
    yarn config set prefix C:\
    yarn global add @esy-nightly/esy@0.6.8-bbdfb8
  displayName: "Install esy (via yarn)"
  condition: eq(variables['AGENT.OS'], 'Windows_NT')

- bash: |
    YARN_BIN_DIR=$(yarn global bin)
    echo $YARN_BIN_DIR
    echo "##vso[task.setvariable variable=YARN_BIN_DIR]$YARN_BIN_DIR"
  displayName: "Setting $YARN_BIN_DIR"

- bash: |
    echo "##vso[task.setvariable variable=PATH]$PATH:$YARN_BIN_DIR" 
  displayName: "Ensure yarn install global tools are available on $PATH"
  condition: ne(variables['AGENT.OS'], 'Windows_NT')

- powershell: Write-Host "##vso[task.setvariable variable=PATH;]${env:PATH};${env:YARN_BIN_DIR}"
  displayName: "Ensure yarn install global tools are available on $PATH"
  condition: eq(variables['AGENT.OS'], 'Windows_NT')
