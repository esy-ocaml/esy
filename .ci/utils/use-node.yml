# steps to use node on agent

steps:
  - task: NodeTool@0
    displayName: "Use Node 15.x"
    inputs:
      versionSpec: 15.x
