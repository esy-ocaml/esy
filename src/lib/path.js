/**
 * @flow
 */

import invariant from 'invariant';
const path = require('path');

/**
 * Absolute path.
 */
export opaque type AbsolutePath: ConcretePath = string;

/**
 * Path with no `.` or `..` or any bash variables in it.
 */
export opaque type ConcretePath: Path = string;

/**
 * Path which can contain bash/make variables in it, cannot be used in
 * filesystem operations.
 */
export opaque type AbstractPath: Path = string;

export opaque type Path: string = AbsolutePath | ConcretePath | AbstractPath;

export function absolute(p: string): AbsolutePath {
  p = sanitizeAbsolutePath(p);
  return p;
}

export function concrete(p: string): ConcretePath {
  p = sanitizeConcretePath(p);
  return p;
}

export function abstract(p: string): AbstractPath {
  return p;
}

export function join<B: Path | string>(base: B, ...p: string[]): B {
  return (path.join(base, ...p): any);
}

export function basename<P: Path | string>(p: P): string {
  return path.basename(p);
}

export function dirname<P: Path | string>(p: P): P {
  return (path.dirname(p): any);
}

export function length(p: ConcretePath): number {
  return p.length;
}

export function toString(p: Path | string): string {
  return p;
}

export const sep: typeof path.sep = path.sep;

export const relative: typeof path.relative = path.relative;

function sanitizeAbsolutePath(p) {
  p = sanitizeConcretePath(p);
  invariant(path.isAbsolute(p), 'Should be absolute path but got: %s', p);
  return p;
}

function sanitizeConcretePath(p) {
  // NOTE: regexpes are stateful, this is why we are creating them here.
  // Consider resetting their state after matching instead.
  const REMOVE_TRAILING_SLASH_RE = /\/+$/g;
  const ENVIRONMENT_VAR_RE = /\$[a-zA-Z_]/g;
  const DOT_PATH_SEGMENT_RE = /\/\.\//g;
  const DOT_DOT_PATH_SEGMENT_RE = /\/\.\.\//g;
  const SANITIZE_SLASH_RE = /\/+/g;

  invariant(
    DOT_PATH_SEGMENT_RE.exec(p) == null,
    'Should not contain "." segments: %s',
    p,
  );
  invariant(
    DOT_DOT_PATH_SEGMENT_RE.exec(p) == null,
    'Should not contain ".." segments: %s',
    p,
  );
  invariant(
    ENVIRONMENT_VAR_RE.exec(p) == null,
    'Path should not contain variable references: %s',
    p,
  );
  p = path.normalize(p);
  p = p.replace(REMOVE_TRAILING_SLASH_RE, '');
  p = p === '' ? '/' : p;
  return p;
}
