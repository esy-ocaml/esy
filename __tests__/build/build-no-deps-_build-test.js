/**
 * @flow
 */

import * as path from 'path';
import {defineTestCaseWithShell} from '../utils';

defineTestCaseWithShell(
  path.join(__dirname, 'fixtures', 'no-deps-_build'),
  `
    run esy build
    run ls
    run ls _build
    run esy x no-deps-_build
    assertStdout "esy x no-deps-_build" "no-deps-_build"
  `,
);
