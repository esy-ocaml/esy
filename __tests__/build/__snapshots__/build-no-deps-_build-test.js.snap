// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build no-deps-_build 1`] = `
"RUNNING: esy config ls
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
RUNNING: esy build
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin needRebuildTarget <sandbox>/node_modules/.cache/_esy/build/bin/build
esy:bin trying to acquire lock
esy:bin:esyBuild execute
esy:bin trying to acquire lock
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:build-dependencies checking if dependencies are built
RUNNING: ls
RUNNING: ls _build
RUNNING: esy x no-deps-_build
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin needRebuildTarget <sandbox>/node_modules/.cache/_esy/build/bin/build
esy:bin trying to acquire lock
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:build-dependencies checking if dependencies are built
RUNNING: esy x no-deps-_build
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:bin needRebuildTarget <sandbox>/node_modules/.cache/_esy/build/bin/build
esy:bin trying to acquire lock
esy:bin no .esyrc found
esy:bin prefix: <testRoot>/esy
esy:build-dependencies checking if dependencies are built"
`;
